block-node/block-access/src/main/java/org/hiero/block/node/access/service/BlockAccessServicePlugin.java

```diff
@@ -0,0 +1,198 @@
+    private Counter requestCounterNotFound;
+
+    /**
+     * Handle a request for a single block
+     *
+     * @param request the request containing the block number or latest flag
+     * @return the response containing the block or an error status
+     */
+    private SingleBlockResponse handleSingleBlockRequest(SingleBlockRequest request) {
+        LOGGER.log(
+                System.Logger.Level.DEBUG, "Received SingleBlockRequest for block number: {0}", request.blockNumber());
+        requestCounter.increment();
+
+        try {
+            // Validate request
+            if (request.retrieveLatest() && request.blockNumber() != 0) {
+                // LOGGER.log(System.Logger.Level.WARNING, "Invalid request: both block_number and retrieve_latest
+                // set");
+                // requestCounterNotFound.increment();
+                // return new SingleBlockResponse(SingleBlockResponseCode.READ_BLOCK_NOT_FOUND, null);
+
+                LOGGER.log(
+                        System.Logger.Level.WARNING,
+                        "Both block_number and retrieve_latest set. Using retrieve_latest instead of block_number: {0}",
+                        request.blockNumber());
+            }
+            long blockNumberToRetrieve;
+
+            // if retrieveLatest is set, get the latest block number
+            if (request.retrieveLatest()) {
+                blockNumberToRetrieve = blockProvider.latestBlockNumber();
+                if (blockNumberToRetrieve < 0) {
+                    LOGGER.log(System.Logger.Level.WARNING, "Latest block number not available");
+                    requestCounterNotAvailable.increment();
+                    return new SingleBlockResponse(SingleBlockResponseCode.READ_BLOCK_NOT_AVAILABLE, null);
+                }
+            } else {
+                blockNumberToRetrieve = request.blockNumber();
+            }
+
+            // Check if block is within available range
+            long lowestBlockNumber = blockProvider.oldestBlockNumber();
+            long highestBlockNumber = blockProvider.latestBlockNumber();
+
+            if (blockNumberToRetrieve < lowestBlockNumber || blockNumberToRetrieve > highestBlockNumber) {
+                LOGGER.log(
+                        System.Logger.Level.WARNING,
+                        "Requested block {0} is outside available range [{1}, {2}]",
+                        blockNumberToRetrieve,
+                        lowestBlockNumber,
+                        highestBlockNumber);
+                requestCounterNotAvailable.increment();
+                return new SingleBlockResponse(SingleBlockResponseCode.READ_BLOCK_NOT_AVAILABLE, null);
+            }
+
+            // Retrieve the block
+            Block block = blockProvider.block(blockNumberToRetrieve).block();
+            requestCounterSuccess.increment();
+            return new SingleBlockResponse(SingleBlockResponseCode.READ_BLOCK_SUCCESS, block);
+
+        } catch (Exception e) {
```

Comment:
It would be good to be _specific_ for all checked exceptions, and then include only `RuntimeException` for "generic" catch.
Example assuming the only checked exception is an `IOException`.
```suggestion
        } catch (IOException | RuntimeException e) {
```

Reply:
Agreed 👍 

✅ Done